tests/acceptance/test_acceptance/test_send_odp_event.py

```diff
@@ -0,0 +1,210 @@
+    :param expected_response:
+    :param expected_status_code:
+    """
+
+    payload = {
+        "type": "test_type",
+        "data": {
+            "idempotence_id": "abc-1234",
+            "data_source_type": "agent",
+        },
+        "identifiers": {"user_id": "test_user_1"},
+        "action": "test_action",
+    }
+
+    resp = create_and_validate_request_and_response(ENDPOINT_SEND_ODP_EVENT, "post", session_override_sdk_key_odp, payload=json.dumps(payload),
+                                                    params={})
+
+    assert json.loads(json.dumps(expected_response)) == resp.json()
+    assert resp.status_code == expected_status_code, resp.text
+    resp.raise_for_status()
+
+
+@pytest.mark.parametrize(
+    "identifiers, action, expected_response, expected_status_code, expected_error, bypass_validation_request, bypass_validation_response", [
+        ({}, "test_action", expected_missing_identifiers, 400,
+         '{"error":"missing or empty \\"identifiers\\" in request payload"}\n', False, False),
+        ("", "test_action", expected_request_error, 400,
+         '{"error":"error parsing request body"}\n', True, True),
+        ({"user_id": "test_user_1"}, "", expected_missing_action,
+         400, '{"error":"missing \\"action\\" in request payload"}\n', False, False)
+    ])
+def test_send_odp_event_invalid_parameters(session_override_sdk_key_odp, identifiers, action, expected_response, expected_status_code, expected_error, bypass_validation_request,
+                                           bypass_validation_response):
+    """
+    Test validates:
+    Returns error when empty identifiers or action is provided.
+    ...
+    :param session_override_sdk_key_odp: sdk key to override the session using odp key
+    :param identifiers:
+    :param action:
+    :param expected_response:
+    :param expected_status_code:
+    :param expected_error
+    """
+
+    payload = {
+        "type": "test_type",
+        "data": {
+            "idempotence_id": "abc-1234",
+            "data_source_type": "agent",
+        },
+        "identifiers": identifiers,
+        "action": action,
+    }
+
+    with pytest.raises(requests.exceptions.HTTPError):
+        resp = create_and_validate_request_and_response(ENDPOINT_SEND_ODP_EVENT, "post", session_override_sdk_key_odp, bypass_validation_request,
+                                                        bypass_validation_response, payload=json.dumps(
+                                                            payload),
+                                                        params={})
+
+        assert json.loads(json.dumps(expected_response)) == resp.json()
+        assert resp.status_code == expected_status_code
+        assert resp.text == expected_error
+        resp.raise_for_status()
+
+
+@pytest.mark.parametrize(
+    "expected_response, expected_status_code, expected_error, bypass_validation_request, bypass_validation_response", [
+        (expected_invalid_data, 500,
```

Comment:
Do we return 500 (not 400) for this user error?

Reply:
yes, in case an error is returned from the sdk side, we return 500.

Reply:
https://github.com/optimizely/agent/blob/ce3b3a708a84d844dc3548b866f973317c3276c8/pkg/handlers/send_odp_event.go#L47

Reply:
[same for ](https://github.com/optimizely/agent/blob/ce3b3a708a84d844dc3548b866f973317c3276c8/pkg/handlers/track.go#L65)

Reply:
> It makes sense. But it's a kind of user request errors. Wondering how we handle other errors when go-sdk (not agent) detects any error in api parameters (like invalid attributes for `decide` api). Do we return 500?

if we validate parameters before executing request, we send 400, but after request execution we return 500. In this case we are not validating separately if `data` only had types that we support, we are counting on go-sdk to do that.